ganttproject/src/main/java/net/sourceforge/ganttproject/storage/SqlProjectDatabaseImpl.kt

```diff
@@ -127,37 +154,58 @@ class SqlProjectDatabaseImpl(private val dataSource: DataSource) : ProjectDataba
+  override fun startTransaction() {
+    if (currentTxn != null) throw ProjectDatabaseException("Previous transaction not committed")
+    currentTxn = mutableListOf()
+  }
+
+  @Throws(ProjectDatabaseException::class)
+  override fun commitTransaction() {
+    try {
+      executeAndLog(currentTxn ?: throw ProjectDatabaseException("No transaction started"))
+    } finally {
+      currentTxn = null
+    }
+  }
+
+  @Throws(ProjectDatabaseException::class)
+  override fun fetchTransactions(startTxnId: Int, limit: Int): List<XlogRecord> = withDSL({ "Failed to fetch log records" }) { dsl ->
     dsl
       .selectFrom(LOGRECORD)
-      .where(LOGRECORD.ID.ge(startId))
-      .orderBy(LOGRECORD.ID.asc())
-      .limit(limit)
-      .map { LogRecord(it.id, it.sqlStatement) }
+      .where(LOGRECORD.LOCAL_TXN_ID.ge(startTxnId).and(LOGRECORD.LOCAL_TXN_ID.lt(startTxnId + limit)))
+      .orderBy(LOGRECORD.LOCAL_TXN_ID, LOGRECORD.ID)
+      .fetchGroups(LOGRECORD.LOCAL_TXN_ID, LOGRECORD.SQL_STATEMENT)
+      .values
+      .map { XlogRecord(it) }
   }
 
-  /** Execute update query and save its xlog. */
+  /** Add update query and save its xlog in the current transaction. */
   @Throws(ProjectDatabaseException::class)
-  internal fun executeUpdate(query: String) = withLog({ "Failed to execute update" }) { query }
+  internal fun update(h2Query: String, postgresQuery: String) = withLog(h2Query, postgresQuery) { "Failed to execute update" }
 }
 
 
 class SqlTaskUpdateBuilder(private val task: Task,
-                           private val executeQuery: (String) -> Unit): TaskUpdateBuilder {
-  private var lastSetStep: UpdateSetMoreStep<TaskRecord>? = null
+                           private val onCommit: (String, String) -> Unit): TaskUpdateBuilder {
+  private var lastSetStepH2: UpdateSetMoreStep<TaskRecord>? = null
+  private var lastSetStepPostgres: UpdateSetMoreStep<TaskRecord>? = null
 
   private fun nextStep(step: (lastStep: UpdateSetStep<TaskRecord>) -> UpdateSetMoreStep<TaskRecord>) {
-    lastSetStep = step(lastSetStep ?: DSL.using(SQL_PROJECT_DATABASE_DIALECT).update(TASK))
+    lastSetStepH2 = step(lastSetStepH2 ?: DSL.using(SQLDialect.H2).update(TASK))
+    lastSetStepPostgres = step(lastSetStepPostgres ?: DSL.using(SQLDialect.POSTGRES).update(TASK))
   }
 
   @Throws(ProjectDatabaseException::class)
-  override fun execute() {
+  override fun commit() {
     try {
-      lastSetStep?.let { updateQuery ->
-        executeQuery(updateQuery.where(TASK.UID.eq(task.uid)).getSQL(ParamType.INLINED))
+      lastSetStepH2?.let { updateQueryH2 ->
+        onCommit(updateQueryH2.where(TASK.UID.eq(task.uid)).getSQL(ParamType.INLINED),
+          lastSetStepPostgres?.where(TASK.UID.eq(task.uid))?.getSQL(ParamType.INLINED)
+            ?: error("No update statement for postgres"))
       }
     } finally {
-      lastSetStep = null
+      lastSetStepH2 = null
```

Comment:
is this builder supposed to be used after commit? If not, you can remove this nullification